PyDSS/controllers.py

```diff
@@ -0,0 +1,333 @@
+
+import abc
+import json
+import logging
+from collections import defaultdict
+from pathlib import Path
+from typing import Optional, List, Union
+
+import opendssdirect as dss
+from pydantic import BaseModel, Field, validator
+
+from PyDSS.SolveMode import get_solver_from_simulation_type
+from PyDSS.common import SimulationType
+from PyDSS.dssInstance import OpenDSS, CONTROLLER_PRIORITIES
+from PyDSS.exceptions import InvalidConfiguration
+from PyDSS.modes.solver_base import SolverBase
+from PyDSS.pyControllers.Controllers.PvController import PvController
+from PyDSS.pyControllers.Controllers.StorageController import StorageController
+from PyDSS.simulation_input_models import ProjectModel
+from PyDSS.utils.dss_utils import iter_elements, list_element_names_by_class
+from PyDSS.utils.timing_utils import TimerStatsCollector, Timer
+from PyDSS.utils.utils import load_data
+
+
+logger = logging.getLogger(__name__)
+
+
+class ControllerBaseModel(BaseModel, abc.ABC):
+    class Config:
+        title = "ControllerBaseModel"
+        anystr_strip_whitespace = True
+        validate_assignment = True
+        validate_all = True
+        extra = "forbid"
+        use_enum_values = False
+        allow_population_by_field_name = True
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_controller_class():
+        """Return the PyDSS controller class for the model."""
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_element_class():
+        """Return the OpenDSS element class for the controller."""
+
+
+class PvControllerModel(ControllerBaseModel):
+    control1: str = Field(
+        title="Control1",
+        description="TODO",
+        alias="Control1",
```

Comment:
The field names are different than what is defined in the current .toml files. But because of the `alias` field, this model is compatible with those files.